NitroxServer/GameLogic/Bases/BaseManager.cs

```diff
@@ -9,153 +9,169 @@ namespace NitroxServer.GameLogic.Bases
 {
     public class BaseManager
     {
-        private Dictionary<NitroxId, BasePiece> partiallyConstructedPiecesById = new Dictionary<NitroxId, BasePiece>();
-        private List<BasePiece> completedBasePieceHistory;
+        // List of all BasePieces
+        // Separation in finished and unfinished Pieces is not needed on server, because the information is on every single object and
+        // the building layout is handled by the client on initial loading.  
+        private List<BasePiece> allBasePieces;
 
-        public BaseManager(List<BasePiece> partiallyConstructedPieces, List<BasePiece> completedBasePieceHistory)
+        public BaseManager(List<BasePiece> allBasePieces)
         {
-            this.completedBasePieceHistory = completedBasePieceHistory;
-            partiallyConstructedPiecesById = partiallyConstructedPieces.ToDictionary(piece => piece.Id);
+            this.allBasePieces = allBasePieces;
         }
 
-        public List<BasePiece> GetCompletedBasePieceHistory()
+        public List<BasePiece> GetAllBasePieces()
         {
-            lock(completedBasePieceHistory)
+            lock (allBasePieces)
             {
-                return new List<BasePiece>(completedBasePieceHistory);
+                return new List<BasePiece>(allBasePieces);
             }
         }
 
-        public List<BasePiece> GetPartiallyConstructedPieces()
+        public void BasePieceConstructionBegin(BasePiece basePiece)
         {
-            lock (partiallyConstructedPiecesById)
-            {
-                return new List<BasePiece>(partiallyConstructedPiecesById.Values);
-            }
-        }
 
-        public void AddBasePiece(BasePiece basePiece)
-        {
-            lock (partiallyConstructedPiecesById)
+#if TRACE && BUILDING
+            NitroxModel.Logger.Log.Debug("BasePieceConstructionBegin - id: " + basePiece.Id + " - basePiece: " + basePiece);
+#endif
+
+            lock (allBasePieces)
             {
-                partiallyConstructedPiecesById.Add(basePiece.Id, basePiece);
+                // Subnautica changes the order of basepieces if there are Layout-Changes on an existing base, e.g. by adding pieces or splitting a base apart
+                // in two bases by removing connector-tubes. This changes the order in which BasePieces are need to be generated by Nitrox on client side loading.
+                if (allBasePieces.Count > 0)
+                {
+                    int nextBuildIndex = allBasePieces.Max(piece => piece.BuildIndex) + 1;
```

Comment:
Might be more friendly to compute upfront in the constructor once and use a member variable to increment.

Reply:
Yes this might be possible. This wouldn't decrement the buildindex when the last built basepiece gots deconstructed, but i see no trouble with missing numbers in the sequence. I will change it